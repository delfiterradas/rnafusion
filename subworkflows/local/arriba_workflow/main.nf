include { ARRIBA_ARRIBA                               } from '../../../modules/nf-core/arriba/arriba/main'
include { SAMTOOLS_INDEX as SAMTOOLS_INDEX_FOR_ARRIBA } from '../../../modules/nf-core/samtools/index/main'
include { SAMTOOLS_SORT as SAMTOOLS_SORT_FOR_ARRIBA   } from '../../../modules/nf-core/samtools/sort/main'
include { SAMTOOLS_VIEW as SAMTOOLS_VIEW_FOR_ARRIBA   } from '../../../modules/nf-core/samtools/view/main'
include { STAR_ALIGN as STAR_FOR_ARRIBA               } from '../../../modules/nf-core/star/align/main'

include { CTATSPLICING_WORKFLOW         }   from '../ctatsplicing_workflow'

workflow ARRIBA_WORKFLOW {
    take:
        reads                           // channel [ meta, [ fastqs ]         ]
        ch_gtf                          // channel [ meta, path_gtf           ]
        ch_fasta                        // channel [ meta, path_fasta         ]
        ch_starindex_ref                // channel [ meta, path_index         ]
        ch_arriba_ref_blacklist         // channel [ meta, path_blacklist     ]
        ch_arriba_ref_cytobands         // channel [ meta, path_cytobands     ]
        ch_arriba_ref_known_fusions     // channel [ meta, path_known_fusions ]
        ch_arriba_ref_protein_domains   // channel [ meta, path_proteins      ]
        ch_starfusion_ref               // channel [ meta, path_starfusion_ref ]
        arriba                          // boolean
        all                             // boolean
        fusioninspector_only            // boolean
        star_ignore_sjdbgtf             // boolean
        ctatsplicing                    // boolean
        seq_center                      // string
        arriba_fusions                  // path
        cram                            // array

    main:

        def ch_versions   = Channel.empty()
        def ch_cram_index = Channel.empty()
        def ch_dummy_file = file("$projectDir/assets/dummy_file_arriba.txt", checkIfExists: true)

        if (( arriba || all ) && !fusioninspector_only) {

            STAR_FOR_ARRIBA(
                reads,
                ch_starindex_ref,
                ch_gtf,
                star_ignore_sjdbgtf,
                '',
                seq_center
            )
            ch_versions = ch_versions.mix(STAR_FOR_ARRIBA.out.versions)

            if ( ctatsplicing || all ) {
                CTATSPLICING_WORKFLOW(
                    STAR_FOR_ARRIBA.out.spl_junc_tab,
                    STAR_FOR_ARRIBA.out.junction,
                    STAR_FOR_ARRIBA.out.bam,
                    ch_starfusion_ref
                )
                ch_versions = ch_versions.mix(CTATSPLICING_WORKFLOW.out.versions)
            }

            if ( arriba_fusions ) {

                ch_arriba_fusions = reads.combine( Channel.value( file( arriba_fusions, checkIfExists: true ) ) )
                    .map { it -> [ it[0], it[2] ] }
                ch_arriba_fusion_fail = ch_dummy_file

            } else {

                ARRIBA_ARRIBA (
                    STAR_FOR_ARRIBA.out.bam,
                    ch_fasta,
                    ch_gtf,
                    ch_arriba_ref_blacklist,
                    ch_arriba_ref_known_fusions,
                    ch_arriba_ref_cytobands,
                    ch_arriba_ref_protein_domains
                )

                ch_versions = ch_versions.mix(ARRIBA_ARRIBA.out.versions)

                ch_arriba_fusions     = ARRIBA_ARRIBA.out.fusions
                ch_arriba_fusion_fail = ARRIBA_ARRIBA.out.fusions_fail.map{ it -> return it[1] }
            }

            if ( cram.contains('arriba') ) {

                SAMTOOLS_SORT_FOR_ARRIBA(STAR_FOR_ARRIBA.out.bam, ch_fasta)
                ch_versions = ch_versions.mix(SAMTOOLS_SORT_FOR_ARRIBA.out.versions )

                SAMTOOLS_VIEW_FOR_ARRIBA(SAMTOOLS_SORT_FOR_ARRIBA.out.bam.map { meta, bam -> [ meta, bam, [] ] }, ch_fasta, [])
                ch_versions = ch_versions.mix(SAMTOOLS_VIEW_FOR_ARRIBA.out.versions )

                SAMTOOLS_INDEX_FOR_ARRIBA(SAMTOOLS_VIEW_FOR_ARRIBA.out.cram)
                ch_versions = ch_versions.mix(SAMTOOLS_INDEX_FOR_ARRIBA.out.versions )

                // Join cram and index files
                ch_cram_index = SAMTOOLS_VIEW_FOR_ARRIBA.out.cram.join(SAMTOOLS_INDEX_FOR_ARRIBA.out.crai)
            }

        } else {

            ch_arriba_fusions = reads
                .combine(Channel.value( file(ch_dummy_file, checkIfExists: true ) ) )
                .map { it -> [ it[0], it[2] ] }

            ch_arriba_fusion_fail = ch_dummy_file
        }

    emit:
        fusions      = ch_arriba_fusions        // channel [ meta, path_fusions   ]
        fusions_fail = ch_arriba_fusion_fail    // channel [ path, fusions_failed ]
        cram_index   = ch_cram_index            // channel [ meta, cram, crai     ]
        versions     = ch_versions              // channel [ versions             ]
    }

