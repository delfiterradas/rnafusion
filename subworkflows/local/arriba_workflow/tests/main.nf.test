nextflow_workflow {

    name "Test Subworkflow ARRIBA_WORKFLOW"
    script "../main.nf"
    workflow "ARRIBA_WORKFLOW"
    tag "subworkflow"
    tag "arriba"
    tag "arriba/arriba"
    tag "samtools"
    tag "samtools/index"
    tag "samtools/sort"
    tag "samtools/view"
    tag "star"
    tag "star/genomegenerate"
    tag "star/align"


    // Test #1 Indexing
    test("ARRIBA_WORKFLOW - Homo sapiens - FASTQs chr4") {

        setup {
            // Create genome index for STAR
            run("STAR_GENOMEGENERATE") {
                script "../../../../modules/nf-core/star/genomegenerate/main.nf"
                process {
                """
                    // FASTA
                    input[0] = Channel.fromPath(
                            "https://raw.githubusercontent.com/nf-core/test-datasets/refs/heads/rnafusion/testdata/reference/ensembl/Homo_sapiens.GRCh38.102.chr4.1700000-54900000.fa", checkIfExists: true
                        )
                        .map{ [[id: it.getName() ], it ]}

                    // GTF
                    input[1] = Channel.fromPath(
                            "https://raw.githubusercontent.com/nf-core/test-datasets/refs/heads/rnafusion/testdata/reference/ensembl/Homo_sapiens.GRCh38.102.chr4.gtf", checkIfExists: true
                        )
                        .map{ [[id: it.getName() ], it ]}
                    """
                }
            }
        }

        when {
            workflow {
                """
                // ch_reads
                input[0] = Channel.of(
                    [
                        [ id: "test_fastqs" ],
                        [
                            file("https://raw.githubusercontent.com/nf-core/test-datasets/refs/heads/rnafusion/testdata/human/reads_1.fq.gz", checkIfExists: true),
                            file("https://raw.githubusercontent.com/nf-core/test-datasets/refs/heads/rnafusion/testdata/human/reads_2.fq.gz", checkIfExists: true)
                        ]
                    ] )

                // ch_gtf
                input[1] =
                    Channel.fromPath(
                        "https://raw.githubusercontent.com/nf-core/test-datasets/refs/heads/rnafusion/testdata/reference/ensembl/Homo_sapiens.GRCh38.102.chr4.gtf", checkIfExists: true
                    )
                    .map{ [ [ id: it.name ], it ] }

                // ch_fasta
                input[2] =
                    Channel.fromPath(
                        "https://raw.githubusercontent.com/nf-core/test-datasets/refs/heads/rnafusion/testdata/reference/ensembl/Homo_sapiens.GRCh38.102.chr4.1700000-54900000.fa", checkIfExists: true
                    )
                    .map{ [ [ id: it.name ], it ] }

                // ch_starindex_ref
                input[3] = STAR_GENOMEGENERATE.out.index

                // ch_arriba_ref_blacklist
                input[4] =
                    Channel.fromPath(
                        "https://raw.githubusercontent.com/nf-core/test-datasets/refs/heads/rnafusion/testdata/reference/arriba/blacklist_hg38_GRCh38_v2.4.0.tsv.gz", checkIfExists: true
                    )
                    .map{ [ [ id: it.name ], it ] }

                // ch_arriba_ref_known_fusions
                input[5] =
                    Channel.fromPath(
                        "https://raw.githubusercontent.com/nf-core/test-datasets/refs/heads/rnafusion/testdata/reference/arriba/known_fusions_hg38_GRCh38_v2.4.0.tsv.gz", checkIfExists: true
                    )
                    .map{ [ [ id: it.name ], it ] }

                // ch_arriba_ref_cytobands
                input[6] =
                    Channel.fromPath(
                        "https://raw.githubusercontent.com/nf-core/test-datasets/refs/heads/rnafusion/testdata/reference/arriba/cytobands_hg38_GRCh38_v2.4.0.tsv", checkIfExists: true
                    )
                    .map{ [ [ id: it.name ], it ] }

                // ch_arriba_ref_protein_domains
                input[7] =
                    Channel.fromPath(
                        "https://raw.githubusercontent.com/nf-core/test-datasets/refs/heads/rnafusion/testdata/reference/arriba/protein_domains_hg38_GRCh38_v2.4.0.gff3", checkIfExists: true
                    )
                    .map{ [ [ id: it.name ], it ] }

                // arriba (boolean)
                input[8] = true

                // all (boolean)
                input[9] = true

                // fusioninspector_only  (boolean)
                input[10] = false

                // star_ignore_sjdbgtf (boolean)
                input[11] = false

                // seq_center (string)
                input[12] = 'test_center'

                // arriba_fusions (path)
                input[13] = null

                // cram (array)
                input[14] = [ 'arriba' ]
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { with(workflow.out) {
                    assert snapshot(
                            file(fusions[0][1]),
                            file(fusions_fail[0]),
                            file(cram_index[0][1]).name,
                            file(cram_index[0][2]).name,
                            versions.collect{ file(it) }
                        ).match()
                    }
                }
            )
        }
    }


    // Test #2 With arriba_fusions file
    test("ARRIBA_WORKFLOW - Homo sapiens - FASTQs chr4 - External fusion file") {

        setup {
            // Create genome index for STAR
            run("STAR_GENOMEGENERATE") {
                script "../../../../modules/nf-core/star/genomegenerate/main.nf"
                process {
                """
                    // FASTA
                    input[0] = Channel.fromPath(
                            "https://raw.githubusercontent.com/nf-core/test-datasets/refs/heads/rnafusion/testdata/reference/ensembl/Homo_sapiens.GRCh38.102.chr4.1700000-54900000.fa", checkIfExists: true
                        )
                        .map{ [[id: it.getName() ], it ]}

                    // GTF
                    input[1] = Channel.fromPath(
                            "https://raw.githubusercontent.com/nf-core/test-datasets/refs/heads/rnafusion/testdata/reference/ensembl/Homo_sapiens.GRCh38.102.chr4.gtf", checkIfExists: true
                        )
                        .map{ [[id: it.getName() ], it ]}
                    """
                }
            }
        }

        when {
            workflow {
                """
                // ch_reads
                input[0] = Channel.of(
                    [
                        [ id: "test_fastqs" ],
                        [
                            file("https://raw.githubusercontent.com/nf-core/test-datasets/refs/heads/rnafusion/testdata/human/reads_1.fq.gz", checkIfExists: true),
                            file("https://raw.githubusercontent.com/nf-core/test-datasets/refs/heads/rnafusion/testdata/human/reads_2.fq.gz", checkIfExists: true)
                        ]
                    ] )

                // ch_gtf
                input[1] =
                    Channel.fromPath(
                        "https://raw.githubusercontent.com/nf-core/test-datasets/refs/heads/rnafusion/testdata/reference/ensembl/Homo_sapiens.GRCh38.102.chr4.gtf", checkIfExists: true
                    )
                    .map{ [ [ id: it.name ], it ] }

                // ch_fasta
                input[2] =
                    Channel.fromPath(
                        "https://raw.githubusercontent.com/nf-core/test-datasets/refs/heads/rnafusion/testdata/reference/ensembl/Homo_sapiens.GRCh38.102.chr4.1700000-54900000.fa", checkIfExists: true
                    )
                    .map{ [ [ id: it.name ], it ] }

                // ch_starindex_ref
                input[3] = STAR_GENOMEGENERATE.out.index

                // ch_arriba_ref_blacklist
                input[4] =
                    Channel.fromPath(
                        "https://raw.githubusercontent.com/nf-core/test-datasets/refs/heads/rnafusion/testdata/reference/arriba/blacklist_hg38_GRCh38_v2.4.0.tsv.gz", checkIfExists: true
                    )
                    .map{ [ [ id: it.name ], it ] }

                // ch_arriba_ref_known_fusions
                input[5] =
                    Channel.fromPath(
                        "https://raw.githubusercontent.com/nf-core/test-datasets/refs/heads/rnafusion/testdata/reference/arriba/known_fusions_hg38_GRCh38_v2.4.0.tsv.gz", checkIfExists: true
                    )
                    .map{ [ [ id: it.name ], it ] }

                // ch_arriba_ref_cytobands
                input[6] =
                    Channel.fromPath(
                        "https://raw.githubusercontent.com/nf-core/test-datasets/refs/heads/rnafusion/testdata/reference/arriba/cytobands_hg38_GRCh38_v2.4.0.tsv", checkIfExists: true
                    )
                    .map{ [ [ id: it.name ], it ] }

                // ch_arriba_ref_protein_domains
                input[7] =
                    Channel.fromPath(
                        "https://raw.githubusercontent.com/nf-core/test-datasets/refs/heads/rnafusion/testdata/reference/arriba/protein_domains_hg38_GRCh38_v2.4.0.gff3", checkIfExists: true
                    )
                    .map{ [ [ id: it.name ], it ] }

                // arriba (boolean)
                input[8] = true

                // all (boolean)
                input[9] = true

                // fusioninspector_only  (boolean)
                input[10] = false

                // star_ignore_sjdbgtf (boolean)
                input[11] = false

                // seq_center (string)
                input[12] = 'test_center'

                // arriba_fusions (string path)
                input[13] = "https://raw.githubusercontent.com/nf-core/test-datasets/refs/heads/rnafusion/testdata/reference/arriba/test_fastqs.arriba.fusions.tsv"

                // cram (array)
                input[14] = [ 'arriba' ]
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { with(workflow.out) {
                    assert snapshot(
                            fusions[0].size() == 2,
                            fusions_fail.size() == 1,
                            file(cram_index[0][1]).name,
                            file(cram_index[0][2]).name,
                            versions.collect{ file(it) }
                        ).match()
                    }
                }
            )
        }
    }

    // TEST #3 WITHOUT INDEXING
    test("ARRIBA_WORKFLOW - Homo sapiens - FASTQs chr4 - cram = []") {

        setup {
            // Create genome index for STAR
            run("STAR_GENOMEGENERATE") {
                script "../../../../modules/nf-core/star/genomegenerate/main.nf"
                process {
                """
                    // FASTA
                    input[0] = Channel.fromPath(
                            "https://raw.githubusercontent.com/nf-core/test-datasets/refs/heads/rnafusion/testdata/reference/ensembl/Homo_sapiens.GRCh38.102.chr4.1700000-54900000.fa", checkIfExists: true
                        )
                        .map{ [[id: it.getName() ], it ]}

                    // GTF
                    input[1] = Channel.fromPath(
                            "https://raw.githubusercontent.com/nf-core/test-datasets/refs/heads/rnafusion/testdata/reference/ensembl/Homo_sapiens.GRCh38.102.chr4.gtf", checkIfExists: true
                        )
                        .map{ [[id: it.getName() ], it ]}
                    """
                }
            }
        }

        when {
            workflow {
                """
                // ch_reads
                input[0] = Channel.of(
                    [
                        [ id: "test_fastqs" ],
                        [
                            file("https://raw.githubusercontent.com/nf-core/test-datasets/refs/heads/rnafusion/testdata/human/reads_1.fq.gz", checkIfExists: true),
                            file("https://raw.githubusercontent.com/nf-core/test-datasets/refs/heads/rnafusion/testdata/human/reads_2.fq.gz", checkIfExists: true)
                        ]
                    ] )

                // ch_gtf
                input[1] =
                    Channel.fromPath(
                        "https://raw.githubusercontent.com/nf-core/test-datasets/refs/heads/rnafusion/testdata/reference/ensembl/Homo_sapiens.GRCh38.102.chr4.gtf", checkIfExists: true
                    )
                    .map{ [ [ id: it.name ], it ] }

                // ch_fasta
                input[2] =
                    Channel.fromPath(
                        "https://raw.githubusercontent.com/nf-core/test-datasets/refs/heads/rnafusion/testdata/reference/ensembl/Homo_sapiens.GRCh38.102.chr4.1700000-54900000.fa", checkIfExists: true
                    )
                    .map{ [ [ id: it.name ], it ] }

                // ch_starindex_ref
                input[3] = STAR_GENOMEGENERATE.out.index

                // ch_arriba_ref_blacklist
                input[4] =
                    Channel.fromPath(
                        "https://raw.githubusercontent.com/nf-core/test-datasets/refs/heads/rnafusion/testdata/reference/arriba/blacklist_hg38_GRCh38_v2.4.0.tsv.gz", checkIfExists: true
                    )
                    .map{ [ [ id: it.name ], it ] }

                // ch_arriba_ref_known_fusions
                input[5] =
                    Channel.fromPath(
                        "https://raw.githubusercontent.com/nf-core/test-datasets/refs/heads/rnafusion/testdata/reference/arriba/known_fusions_hg38_GRCh38_v2.4.0.tsv.gz", checkIfExists: true
                    )
                    .map{ [ [ id: it.name ], it ] }

                // ch_arriba_ref_cytobands
                input[6] =
                    Channel.fromPath(
                        "https://raw.githubusercontent.com/nf-core/test-datasets/refs/heads/rnafusion/testdata/reference/arriba/cytobands_hg38_GRCh38_v2.4.0.tsv", checkIfExists: true
                    )
                    .map{ [ [ id: it.name ], it ] }

                // ch_arriba_ref_protein_domains
                input[7] =
                    Channel.fromPath(
                        "https://raw.githubusercontent.com/nf-core/test-datasets/refs/heads/rnafusion/testdata/reference/arriba/protein_domains_hg38_GRCh38_v2.4.0.gff3", checkIfExists: true
                    )
                    .map{ [ [ id: it.name ], it ] }

                // arriba (boolean)
                input[8] = true

                // all (boolean)
                input[9] = true

                // fusioninspector_only  (boolean)
                input[10] = false

                // star_ignore_sjdbgtf (boolean)
                input[11] = false

                // seq_center (string)
                input[12] = 'test_center'

                // arriba_fusions (path)
                input[13] = null

                // cram (array)
                input[14] = [ ]
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { with(workflow.out) {
                    assert snapshot(
                            file(fusions[0][1]),
                            file(fusions_fail[0]),
                            cram_index.size() == 0,
                            versions.collect{ file(it) }
                        ).match()
                    }
                }
            )
        }
    }

}
