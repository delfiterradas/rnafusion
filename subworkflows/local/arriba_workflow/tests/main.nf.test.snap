{
    "ARRIBA_WORKFLOW - Homo sapiens - FASTQs chr4 - External fusion file": {
        "content": [
            true,
            true,
            "test_fastqs_star_for_arriba_sorted.cram",
            "test_fastqs_star_for_arriba_sorted.cram.crai",
            [
                "versions.yml:md5,439bbb92ff0a83f1e278fc396e9d8ce9",
                "versions.yml:md5,85458747b55f37c1a5afd39ee7a3a4aa",
                "versions.yml:md5,bfc5d96804f2991c7f7c705f1ddf81ec",
                "versions.yml:md5,f4f64d3f1fd867d5afa51e03f7cf2824"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.10.2"
        },
        "timestamp": "2024-12-11T15:53:59.18258718"
    },
    "ARRIBA_WORKFLOW - Homo sapiens - FASTQs chr4": {
        "content": [
            "test_fastqs.arriba.fusions.tsv:md5,8f39789c4428e81eb9a8d0e54c34c43d",
            "test_fastqs.arriba.fusions.discarded.tsv:md5,b804c1ed5b01d34163f5c0b2f6810f98",
            "test_fastqs_star_for_arriba_sorted.cram",
            "test_fastqs_star_for_arriba_sorted.cram.crai",
            [
                "versions.yml:md5,24030f38976402fad0861e6ec99ee6b6",
                "versions.yml:md5,439bbb92ff0a83f1e278fc396e9d8ce9",
                "versions.yml:md5,85458747b55f37c1a5afd39ee7a3a4aa",
                "versions.yml:md5,bfc5d96804f2991c7f7c705f1ddf81ec",
                "versions.yml:md5,f4f64d3f1fd867d5afa51e03f7cf2824"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.10.2"
        },
        "timestamp": "2024-12-11T15:43:48.053656601"
    },
    "ARRIBA_WORKFLOW - Homo sapiens - FASTQs chr4 - cram = []": {
        "content": [
            "test_fastqs.arriba.fusions.tsv:md5,8f39789c4428e81eb9a8d0e54c34c43d",
            "test_fastqs.arriba.fusions.discarded.tsv:md5,b804c1ed5b01d34163f5c0b2f6810f98",
            true,
            [
                "versions.yml:md5,24030f38976402fad0861e6ec99ee6b6",
                "versions.yml:md5,439bbb92ff0a83f1e278fc396e9d8ce9"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.10.2"
        },
        "timestamp": "2024-12-11T16:07:37.079418154"
    }
}